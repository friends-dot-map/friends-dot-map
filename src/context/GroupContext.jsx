import { createContext, useContext, useState, useEffect } from 'react';
import { getAllProfiles } from '../services/profiles';

export const GroupContext = createContext();

const GroupProvider = ({ children }) => {
  const [group, setGroup] = useState([]);

  useEffect(() => {
    const fetchGroup = async () => {
      try {
        const data = await getAllProfiles();
        setGroup(data);
      } catch (error) {
        throw error('something went horribly wrong');
      }
    };
    fetchGroup();
  }, []);
  const groupValue = { group, setGroup };
  return (
    <GroupContext.Provider value={groupValue}>{children}</GroupContext.Provider>
  );
};

const useGroup = () => {
  const context = useContext(GroupContext);

  if (context === undefined) {
    throw new Error('useGroup must be used within a GroupProvider');
  }

  return context;
};

export { GroupProvider, useGroup };
