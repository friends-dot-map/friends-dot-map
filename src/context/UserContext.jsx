import { createContext, useContext, useState } from 'react';
import { getUser } from '../services/users';

export const UserContext = createContext();

const UserProvider = ({ children }) => {
  const currentUser = getUser();
  const [user, setUser] = useState(
    currentUser ? { id: currentUser.id, email: currentUser.email } : {}
  );

  const userValues = { user, setUser };

  return (
    <UserContext.Provider value={userValues}>{children}</UserContext.Provider>
  );
};

const useUser = () => {
  const context = useContext(UserContext);

  if (context === undefined) {
    throw new Error('useUser must be used within a UserProvider');
  }

  return context;
};

export { UserProvider, useUser };
