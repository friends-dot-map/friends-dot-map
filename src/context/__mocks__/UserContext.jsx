import { createContext, useContext, useState } from 'react';

export const UserContext = createContext();

const UserProvider = ({ mockUser, children }) => {
  const [user, setUser] = useState(
    mockUser ? { id: mockUser.id, email: mockUser.email } : {}
  );

  const userValues = { user, setUser };

  return (
    <UserContext.Provider value={userValues}>{children}</UserContext.Provider>
  );
};

const useUser = () => {
  const context = useContext(UserContext);

  if (context === undefined) {
    throw new Error('useUser must be used within a UserProvider');
  }

  return context;
};

export { UserProvider, useUser };
