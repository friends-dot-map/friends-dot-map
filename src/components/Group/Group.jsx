import { Link } from 'react-router-dom';
import { useGroup } from '../../context/GroupContext';

export default function Group() {
  const { group } = useGroup();

  return (
    <div className="flex flex-col w-3/4 mx-auto space-y-3 items-center">
      <h1 className="text-3xl md:text-5xl font-cursive text-center mt-10">
        Your groups
      </h1>
      <hr className="w-40 md:w-60 h-1 bg-dark/30 border-0 rounded-md" />
      <h2 className="text-2xl text-center">friends.map(🐞)</h2>

      <ul className="flex flex-col w-full space-y-5 p-2">
        {group.map((friend) => (
          <li
            key={friend.id}
            className="flex items-center rounded-md bg-white/30 ring-tint p-3 gap-4"
          >
            <span className="text-4xl md:text-6xl">{friend.avatar}</span>
            <div className="flex flex-col md:flex-row md: items-baseline md:space-x-2">
              <Link
                to={`/profile/${friend.username}`}
                className="font-cursive text-xl md:text-3xl tracking-wide"
              >
                {friend.username}
              </Link>
              <span className="text-xs md:text-lg">({friend.first_name})</span>
            </div>
          </li>
        ))}
      </ul>
    </div>
  );
}
