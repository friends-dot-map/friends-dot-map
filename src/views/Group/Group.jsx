import React from 'react';
import { Link } from 'react-router-dom';
import { useGroup } from '../../context/GroupContext';

export default function Group() {
  const { group } = useGroup();

  return (
    <ul>
      {group.map((friend) => (
        <li key={friend.id}>
          <Link to={`/profile/${friend.username}`}>
            {friend.avatar} {friend.username}
          </Link>
        </li>
      ))}
    </ul>
  );
}
