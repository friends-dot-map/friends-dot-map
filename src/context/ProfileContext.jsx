import { createContext, useContext, useState, useEffect } from 'react';
import { getProfile } from '../services/profiles';
import { useUser } from './UserContext';
import { formatDate } from '../utils/utils';

export const ProfileContext = createContext();

const ProfileProvider = ({ children }) => {
  const [userCoords, setUserCoords] = useState({ latitude: '', longitude: '' });
  const [profile, setProfile] = useState({
    username: '',
    first_name: '',
    likes: '',
    status: '',
    avatar: '',
    coords: userCoords,
    updated_at: formatDate(),
  });
  const [loading, setLoading] = useState(true);
  const { user } = useUser();

  useEffect(() => {
    const fetchProfile = async () => {
      try {
        const profileData = await getProfile(user.email);
        setProfile(profileData);
      } catch (error) {
        setProfile({});
      }
      setLoading(false);
    };
    fetchProfile();
  }, [user]);

  const profileValues = {
    profile,
    loading,
    setLoading,
    setProfile,
    userCoords,
    setUserCoords,
  };

  return (
    <ProfileContext.Provider value={profileValues}>
      {children}
    </ProfileContext.Provider>
  );
};

const useProfile = () => {
  const context = useContext(ProfileContext);

  if (context === undefined) {
    throw new Error('useProfile must be used within a ProfileProvider');
  }

  return context;
};

export { ProfileProvider, useProfile };
