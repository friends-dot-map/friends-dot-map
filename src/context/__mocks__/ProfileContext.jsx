import { createContext, useContext, useState, useEffect } from 'react';
import { useUser } from '../UserContext';
import { formatDate } from '../../utils/utils';

export const ProfileContext = createContext();

const ProfileProvider = ({ mockProfile, children }) => {
  const [userCoords, setUserCoords] = useState({ latitude: '', longitude: '' });
  const [profile, setProfile] = useState({
    username: '',
    first_name: '',
    likes: '',
    status: '',
    avatar: '',
    coords: userCoords,
    updated_at: formatDate(),
  });
  const [profileLoading, setProfileLoading] = useState(true);
  const { user } = useUser();

  useEffect(() => {
    const fetchProfile = async () => {
      try {
        const profileData = mockProfile || {};
        setProfile(profileData);
      } catch (error) {
        setProfile({});
      }
      setProfileLoading(false);
    };
    fetchProfile();
  }, [user]);

  const profileValues = {
    profile,
    profileLoading,
    setProfile,
    userCoords,
    setUserCoords,
  };

  return (
    <ProfileContext.Provider value={profileValues}>
      {children}
    </ProfileContext.Provider>
  );
};

const useProfile = () => {
  const context = useContext(ProfileContext);

  if (context === undefined) {
    throw new Error('useProfile must be used within a ProfileProvider');
  }

  return context;
};

export { ProfileProvider, useProfile };
