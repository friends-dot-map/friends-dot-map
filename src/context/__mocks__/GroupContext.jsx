import { createContext, useContext, useState, useEffect } from 'react';
import { useProfile } from '../ProfileContext'


export const GroupContext = createContext();

const GroupProvider = ({ mockGroup, children }) => {
  const [group, setGroup] = useState([]);
  const { profile } = useProfile();

  useEffect(() => {
    const fetchGroup = async () => {
      try {
        const data = mockGroup || [];
        setGroup(data);
      } catch (error) {
        throw error('something went horribly wrong');
      }
    };
    fetchGroup();
  }, [profile]);

  const groupValue = { group };

  return (
    <GroupContext.Provider value={groupValue}>{children}</GroupContext.Provider>
  );
};

const useGroup = () => {
  const context = useContext(GroupContext);

  if (context === undefined) {
    throw new Error('useGroup must be used within a GroupProvider');
  }

  return context;
};

export { GroupProvider, useGroup };
