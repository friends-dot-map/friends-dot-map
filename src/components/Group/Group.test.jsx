import { render, screen, waitFor } from '@testing-library/react';
import { Route } from 'react-router-dom';
import { MemoryRouter } from 'react-router-dom';
import { GroupProvider } from '../../context/GroupContext';
import { ProfileProvider } from '../../context/ProfileContext';
import { UserProvider } from '../../context/UserContext';

import Group from './Group';

const mockProfile = {
  avatar: '🐶',
  username: 'spongebob',
  first_name: 'charles',
  likes: 'snails',
  status: 'rockin in the free world',
  coords: { latitude: 69, longitude: 420 },
};

const mockUser = {
  id: 33,
  email: 'billclinton@example.com',
  password: 'ilovemonica',
};

const mockGroup = [
  {
    avatar: '🐶',
    username: 'spongebob',
    first_name: 'charles',
    likes: 'snails',
    status: 'rockin in the free world',
    coords: { latitude: 69, longitude: 420 },
  },
];

jest.mock('../../context/ProfileContext');
jest.mock('../../context/UserContext');
jest.mock('../../context/GroupContext');

test('should render the header', () => {
  const { container } = render(
    <UserProvider mockUser={mockUser}>
      <ProfileProvider mockProfile={mockProfile}>
        <GroupProvider mockGroup={mockGroup}>
          <MemoryRouter initialEntries={['/group']}>
            <Route path="/group">
              <Group />
            </Route>
          </MemoryRouter>
        </GroupProvider>
      </ProfileProvider>
    </UserProvider>
  );

  expect(container).toMatchSnapshot();

  return waitFor(() => {
    expect(screen.queryByLabelText(/loader/i)).not.toBeInTheDocument();
  });
});
