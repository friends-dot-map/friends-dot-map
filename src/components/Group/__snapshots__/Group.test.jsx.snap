// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the header 1`] = `
<div>
  <ul>
    <li>
      <a
        href="/profile/spongebob"
      >
        🐶
         
        spongebob
      </a>
    </li>
  </ul>
</div>
`;
